### DISCUSSION TITLE
What issue is a good bug / feature for new Quiver developers?

### DISCUSSION THREAD

**Discussion by enjoysmath** [varkor/quiver#135] [General]
I'm using Quiver locally and loading into an iframe, but there are features of my app, that aren't really relevant to the editor itself - those will be coded by me in JS / Python independently of Quiver, and there are features that should be part of Quiver, so my workflow should be to help update quiver - once the update is on a quiver branch, then update my code with it, apply my patch to call the GUI actions.

Anyway, so I'm here to help improve Quiver editor itself.  If you have an idea of a good starter issue, please reply here.

I also use Quiver on a daily basis for learning category theory, and can see that some additional features are needed.

**Comment by varkor** [comment]
Several issues I think should be fairly easy to address are these:
- https://github.com/varkor/quiver/issues/101
- https://github.com/varkor/quiver/issues/107
- https://github.com/varkor/quiver/issues/110
- https://github.com/varkor/quiver/issues/120

I can offer more specific advice on the issues themselves if needed.

**Reply by enjoysmath** [reply]
Thanks @varkor, I will work on $\pi/2$ flipping most likely.  Some time this week :) 

Questions: 
1. Where would you prefer the button to be located?
2. I think it should work with current selection, is that cool?  If the selection is none, then the whole diagram gets rotated.

**Reply by varkor** [reply]
That sounds great :)
Regarding (1), I imagine it would work best as a toolbar action at the top. Probably it's okay to have one more button added to it. Regarding (2), yes, that sounds like a good strategy.

**Reply by enjoysmath** [reply]
@varkor I created a simple function `centre_point` within quiver that takes the center point of the quiver's bounding rect.  I then call it from this function, but I'm stumped as to how to actually set the the vertex positions, once I've computed their new positions.  See below near the question mark in the code.  I think this function should go in UI class, since Vertex's `set_position` needs a ui parameter.

```
    // Rotate quiver by pi/2 degrees CCW about center_point().
    rotate_90deg_ccw() {
        const center = this.centre_point();
        
        if (center !== null) {
            const vertices = Array.from(this.cells[0]);

            for (let i=0; i < vertices.length; i++)
            {
                const position = vertices[i].position;
                position.set_pos(Vertex())
                // new_position-position.y + center[1] + center[0]; ??
                position.y = position.x - center[0] + center[1];
            }          
        }
    }    

```

Also, calling `ui.quiver.rerender(ui)` after this is causing the grid to get offset where it shouldn't be.

**Reply by varkor** [reply]
You can look at how cells are moved using the arrow keys as a reference for moving cells around: https://github.com/varkor/quiver/blob/master/src/ui.js#L2069-L2102
I don't think you need to find the centre of the diagram: simply the bounding box of the diagram. I think a rotate method would be ideally placed in the `UI` class, I agree.

**Comment by enjoysmath** [comment]
@varkor  do you think you could set me up with your debugging environment?  For example if you use vscode, can you provide me with a launch.json (that's all I need really).  That launches Chrome browser directed at the src/index.html with Javascript debugging enabled?  I'll post mine here if I come across it.  Found one that works:

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch Chrome",
            "request": "launch",
            "type": "pwa-chrome",
            "url": "${workspaceFolder}/src/index.html",
            // "webRoot": ""
        },
    ]
}
```

**Reply by varkor** [reply]
Personally, I just open `index.html` up in my web browser and use the debugging tools there: I don't launch from Visual Studio Code.
